### ISSUE TITLE
Pre-train model on ImageNet model (Or training suggestions)

### ISSUE THREAD

**Issue by yqtian-se** [JiahuiYu/generative_inpainting#389]
Hi,

I hope to use this project on ImageNet dataset. I am wondering to know:
1. Have you used ImageNet to train the model before? Any suggestions on hyperparameters? Do you have pre-trained model? 
2. I noticed that it seems I only need to provide images. Do I need to provide the masks during the training?

Thanks

**Comment by JiahuiYu** [comment]
1. For DeepFill v1, we do have ImageNet pretrained models. For DeepFill v2, I suggest to use the same hyper-parameters. 

2. No. Masks are automatically and randomly generated.

**Comment by yqtian-se** [comment]
Thanks. I will have a try. 

Happy New Year.

**Comment by struggleeachday** [comment]
I am trying to train DeepFiill v2 using ImageNet. The training filst includes all training images of ImageNet 2012 set and filst val includes all validation images. 

I am wondering to know when should I stop the training? I noticed the loss is negative after several steps. 

> train epoch 1, iter 4000/4000, loss 0.540666, 1.61 batches/sec.
train epoch 2, iter 4000/4000, loss 0.283518, 1.61 batches/sec.
train epoch 3, iter 4000/4000, loss 0.122314, 1.61 batches/sec.
train epoch 4, iter 4000/4000, loss 0.006348, 1.61 batches/sec.
train epoch 5, iter 4000/4000, loss 0.025874, 1.61 batches/sec.
train epoch 6, iter 4000/4000, loss 0.017363, 1.61 batches/sec.
train epoch 7, iter 4000/4000, loss 0.005975, 1.61 batches/sec.
train epoch 8, iter 4000/4000, loss -0.005902, 1.61 batches/sec.
train epoch 9, iter 4000/4000, loss -0.012063, 1.61 batches/sec.
train epoch 10, iter 4000/4000, loss -0.017150, 1.61 batches/sec.
train epoch 11, iter 4000/4000, loss -0.020469, 1.61 batches/sec.
train epoch 12, iter 4000/4000, loss -0.023826, 1.61 batches/sec.
train epoch 13, iter 4000/4000, loss -0.033107, 1.61 batches/sec.
train epoch 14, iter 4000/4000, loss -0.037790, 1.62 batches/sec.
train epoch 15, iter 4000/4000, loss -0.036733, 1.62 batches/sec.
train epoch 16, iter 4000/4000, loss -0.044607, 1.62 batches/sec.
train epoch 17, iter 4000/4000, loss -0.043068, 1.62 batches/sec.
train epoch 18, iter 4000/4000, loss -0.045801, 1.62 batches/sec.
train epoch 19, iter 4000/4000, loss -0.053231, 1.62 batches/sec.
train epoch 20, iter 4000/4000, loss -0.063750, 1.62 batches/sec.
train epoch 21, iter 4000/4000, loss -0.067062, 1.61 batches/sec.


Following is my inpainting.yml, I only changed dataset and radom_crop. Any further suggestions?

```
# =========================== Basic Settings ===========================
# machine info
num_gpus_per_job: 1  # number of gpus each job need
num_cpus_per_job: 4  # number of gpus each job need
num_hosts_per_job: 1
memory_per_job: 32  # number of gpus each job need
gpu_type: 'nvidia-tesla-p100'

# parameters
name: places2_gated_conv_v100  # any name
model_restore: ''  # logs/places2_gated_conv
dataset: 'imagenet'  # 'tmnist', 'dtd', 'places2', 'celeba', 'imagenet', 'cityscapes'
random_crop: True  # Set to false when dataset is 'celebahq', meaning only resize the images to img_shapes, instead of crop img_shapes from a larger raw image. This is useful when you train on images with different resolutions like places2. In these cases, please set random_crop to true.
val: False  # true if you want to view validation results in tensorboard
log_dir: logs/full_model_celeba_hq_256

gan: 'sngan'
gan_loss_alpha: 1
gan_with_mask: True
discounted_mask: True
random_seed: False
padding: 'SAME'

# training
train_spe: 4000
max_iters: 100000000
viz_max_out: 10
val_psteps: 2000

# data
data_flist:
  # https://github.com/jiahuiyu/progressive_growing_of_gans_tf
  celebahq: [
    'data/celeba_hq/train_shuffled.flist',
    'data/celeba_hq/validation_static_view.flist'
  ]
  # http://mmlab.ie.cuhk.edu.hk/projects/celeba.html, please to use random_crop: True
  celeba: [
    'data/celeba/train_shuffled.flist',
    'data/celeba/validation_static_view.flist'
  ]
  # http://places2.csail.mit.edu/, please download the high-resolution dataset and use random_crop: True
  places2: [
    'data/places2/train_shuffled.flist',
    'data/places2/validation_static_view.flist'
  ]
  # http://www.image-net.org/, please use random_crop: True
  imagenet: [
    'data/imagenet/train_shuffled.flist',
    'data/imagenet/validation_static_view.flist',
  ]

static_view_size: 30
img_shapes: [256, 256, 3]
height: 128
width: 128
max_delta_height: 32
max_delta_width: 32
batch_size: 16
vertical_margin: 0
horizontal_margin: 0

# loss
ae_loss: True
l1_loss: True
l1_loss_alpha: 1.

# to tune
guided: False
edge_threshold: 0.6
```

**Comment by yqtian-se** [comment]
I am trying to train DeepFiill v2 using ImageNet. The training filst includes all training images of ImageNet 2012 set and filst val includes all validation images.

I am wondering to know when should I stop the training? I noticed the loss is negative after several steps.

> train epoch 1, iter 4000/4000, loss 0.540666, 1.61 batches/sec.
> train epoch 2, iter 4000/4000, loss 0.283518, 1.61 batches/sec.
> train epoch 3, iter 4000/4000, loss 0.122314, 1.61 batches/sec.
> train epoch 4, iter 4000/4000, loss 0.006348, 1.61 batches/sec.
> train epoch 5, iter 4000/4000, loss 0.025874, 1.61 batches/sec.
> train epoch 6, iter 4000/4000, loss 0.017363, 1.61 batches/sec.
> train epoch 7, iter 4000/4000, loss 0.005975, 1.61 batches/sec.
> train epoch 8, iter 4000/4000, loss -0.005902, 1.61 batches/sec.
> train epoch 9, iter 4000/4000, loss -0.012063, 1.61 batches/sec.
> train epoch 10, iter 4000/4000, loss -0.017150, 1.61 batches/sec.
> train epoch 11, iter 4000/4000, loss -0.020469, 1.61 batches/sec.
> train epoch 12, iter 4000/4000, loss -0.023826, 1.61 batches/sec.
> train epoch 13, iter 4000/4000, loss -0.033107, 1.61 batches/sec.
> train epoch 14, iter 4000/4000, loss -0.037790, 1.62 batches/sec.
> train epoch 15, iter 4000/4000, loss -0.036733, 1.62 batches/sec.
> train epoch 16, iter 4000/4000, loss -0.044607, 1.62 batches/sec.
> train epoch 17, iter 4000/4000, loss -0.043068, 1.62 batches/sec.
> train epoch 18, iter 4000/4000, loss -0.045801, 1.62 batches/sec.
> train epoch 19, iter 4000/4000, loss -0.053231, 1.62 batches/sec.
> train epoch 20, iter 4000/4000, loss -0.063750, 1.62 batches/sec.
> train epoch 21, iter 4000/4000, loss -0.067062, 1.61 batches/sec.

Following is my inpainting.yml, I only changed dataset and radom_crop. Any further suggestions?
```
# =========================== Basic Settings ===========================
# machine info
num_gpus_per_job: 1  # number of gpus each job need
num_cpus_per_job: 4  # number of gpus each job need
num_hosts_per_job: 1
memory_per_job: 32  # number of gpus each job need
gpu_type: 'nvidia-tesla-p100'

# parameters
name: places2_gated_conv_v100  # any name
model_restore: ''  # logs/places2_gated_conv
dataset: 'imagenet'  # 'tmnist', 'dtd', 'places2', 'celeba', 'imagenet', 'cityscapes'
random_crop: True  # Set to false when dataset is 'celebahq', meaning only resize the images to img_shapes, instead of crop img_shapes from a larger raw image. This is useful when you train on images with different resolutions like places2. In these cases, please set random_crop to true.
val: False  # true if you want to view validation results in tensorboard
log_dir: logs/full_model_celeba_hq_256

gan: 'sngan'
gan_loss_alpha: 1
gan_with_mask: True
discounted_mask: True
random_seed: False
padding: 'SAME'

# training
train_spe: 4000
max_iters: 100000000
viz_max_out: 10
val_psteps: 2000

# data
data_flist:
  # https://github.com/jiahuiyu/progressive_growing_of_gans_tf
  celebahq: [
    'data/celeba_hq/train_shuffled.flist',
    'data/celeba_hq/validation_static_view.flist'
  ]
  # http://mmlab.ie.cuhk.edu.hk/projects/celeba.html, please to use random_crop: True
  celeba: [
    'data/celeba/train_shuffled.flist',
    'data/celeba/validation_static_view.flist'
  ]
  # http://places2.csail.mit.edu/, please download the high-resolution dataset and use random_crop: True
  places2: [
    'data/places2/train_shuffled.flist',
    'data/places2/validation_static_view.flist'
  ]
  # http://www.image-net.org/, please use random_crop: True
  imagenet: [
    'data/imagenet/train_shuffled.flist',
    'data/imagenet/validation_static_view.flist',
  ]

static_view_size: 30
img_shapes: [256, 256, 3]
height: 128
width: 128
max_delta_height: 32
max_delta_width: 32
batch_size: 16
vertical_margin: 0
horizontal_margin: 0

# loss
ae_loss: True
l1_loss: True
l1_loss_alpha: 1.

# to tune
guided: False
edge_threshold: 0.6
```

**Comment by JiahuiYu** [comment]
Losses in tensorboard could help. Please search in closed issues to see how to interpret losses in tensorboard.

**Comment by yqtian-se** [comment]
Thanks. I learned how to use tensorboard to check the losses. And I understood that the loss in the terminal could be negative. 

Following images are my current results from imagenet. And the training has run more than 4 days (134 epochs). The smoothing value in ternsorboard is 0.8. 

In other issue posts, you mentioned that we should train the model until all losses converged. I think in my results, the aeloss is converging. while the others seem not....  I am wondering to know do you have any suggestions? 

I am also worrried whether the whole imagenet training set is too large? 
Thanks 

![image](https://user-images.githubusercontent.com/41743566/71780813-337b2e00-3002-11ea-8133-62dda581858d.png)

![image](https://user-images.githubusercontent.com/41743566/71780820-3fff8680-3002-11ea-8b07-4fd656bd65f4.png)

**Comment by JiahuiYu** [comment]
You can focus on the AE loss and D loss. In your figures, both are converging so keep training, don’t worry.

**Comment by syomantak** [comment]
@yqtianust did you get it to work? Can you share the pre-trained model?
